src/ring_theory/congruence.lean

```diff
@@ -0,0 +1,259 @@
+/-
+Copyright (c) 2022 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+
+import group_theory.congruence
+import algebra.ring.inj_surj
+import algebra.hom.ring
+
+/-!
+# Congruence relations on rings
+
+This file defines congruence relations on rings, which extend `con` and `add_con` on monoids and
+additive monoids.
+
+Most of the time you likely want to use the `ideal.quotient` API that is built on top of this.
+
+## Main Definitions
+
+* `ring_con R`: the type of congruence relations respecting `+` and `*`.
+* `ring_con_gen r`: the inductively defined smallest ring congruence relation containing a given
+  binary relation.
+
+## TODO
+
+* Use this for `ring_quot` too.
+* Copy across more API from `con` and `add_con` in `group_theory/congruence.lean`.
+
+-/
+
+/-- A congruence relation on a type with an addition and multiplication is an equivalence relation
+which preserves both. -/
+structure ring_con (R : Type*) [has_add R] [has_mul R] extends setoid R :=
+(add' : ∀ {w x y z}, r w x → r y z → r (w + y) (x + z))
+(mul' : ∀ {w x y z}, r w x → r y z → r (w * y) (x * z))
+
+variables {α R : Type*}
+
+/-- The inductively defined smallest ring congruence relation containing a given binary
+    relation. -/
+inductive ring_con_gen.rel [has_add R] [has_mul R] (r : R → R → Prop) : R → R → Prop
+| of : Π x y, r x y → ring_con_gen.rel x y
+| refl : Π x, ring_con_gen.rel x x
+| symm : Π {x y}, ring_con_gen.rel x y → ring_con_gen.rel y x
+| trans : Π {x y z}, ring_con_gen.rel x y → ring_con_gen.rel y z → ring_con_gen.rel x z
+| add : Π {w x y z}, ring_con_gen.rel w x → ring_con_gen.rel y z → ring_con_gen.rel (w + y) (x + z)
+| mul : Π {w x y z}, ring_con_gen.rel w x → ring_con_gen.rel y z → ring_con_gen.rel (w * y) (x * z)
+
+/-- The inductively defined smallest ring congruence relation containing a given binary
+    relation. -/
```

Comment:
Could you prove that this is indeed the smallest congruence relation extending `r`? Or is that too much work, and leave it for a followup?

Reply:
To state that in a way consistent with `mul_con` I would need to provide a `complete_lattice` structure; so I think that's too much work for now.

Reply:
I've extended the TODO to mention that explicitly.

Reply:
Done in #18588